Feat/upload bank accounts

paulopraddo opened the pull request.

paulopraddo committed:
upload bank accounts

paulopraddo committed:
upload Bank accounts

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,73 @@
+public class ValidateUserBankV8Service implements ValidateUserBankV8UseCase {
+
+    @Value("${api.v8.bank.url}")
+    private String urlGetToken;
```
aqui deve ser uma classe Configurações, nao no app nem do Domain, no core da aplicação.
Deve pegar as envs do application.properties e configurar uma provider pra fazer o getUrlV8, e instanciar a dataProvider na service, pode configurar a classe Configuração dentro da Provider, e criar um metodo que vai dar o get, e esse metodo voce puxa dentro da service pra pegar o valor da env.

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,73 @@
+    @Value("${api.v8.bank.url}")
+    private String urlGetToken;
+
+    private final V8BankEnviroment v8BankEnviroment;
```
aqui voce nao instancia dessa maneira, vc faz um V8BankRestModel/DTO e faz um new dentro do metodo execute.
Esse cara nao pode ser uma config, tem que ser uma Restmodel/DTO pra enviar para o httpRequest.

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,73 @@
+            formData.add("scope", v8BankEnviroment.getScope());
+            formData.add("client_id", v8BankEnviroment.getClientId());
+
+            System.out.println("grant_type: " + v8BankEnviroment.getGrantType());
```
Remover os System.out.print do código, nao precisamos de logs, nao iremos utilizar pra nada, clean Code nao tem System out, tem tratamento de exceção

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,73 @@
+
+            
+        } catch (Exception e) {
+            e.printStackTrace();
```
que isso? tem que criar uma exceção  personalizada, conforme clean arch e código bem escrito, NUNCA FACA PRINTSTACKTRACE!!!

Bighetto commented on src/main/java/api/config/V8BankEnviroment.java:
```diff
@@ -0,0 +1,28 @@
+@Setter
+@Data
+@Component
+public class V8BankEnviroment {
```
Tem que ser model isso, ou restModel, classe de Environment nao pode ser assim. Seguir conselho no primeiro comentário

Bighetto commented on src/main/java/api/config/V8BankEnviroment.java:
```diff
@@ -0,0 +1,28 @@
+    @Value("${api.v8.bank.audience}")
+    private String audience;
+
+    @Value("${api.v8.bank.clientid}")
```
Nao pode pegar @value direto na model, configurar método na provider q vai buscar env do Application.properties

Bighetto commented on src/main/java/api/security/auth/domain/dataprovider/AuthDataProvider.java:
```diff
@@ -12,5 +12,7 @@ public interface AuthDataProvider {
 
     void updatePassword(String document, String password);
+
+    UserLogin findUserLoginByUserEmail(String userEmail);
```
Retornar Entity, nao model, lembra da regra, vai entrar na Domain, so pode entrar entity.

Bighetto reviewed (changes requested):
Realizar alterações e enviar para revisão novamente.

paulopraddo committed:
Refac uploadBankUser

paulopraddo committed:
Refac validate user and tests

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,65 @@
+                );
+            
+        } catch (Exception e) {
+            
```
nao tem exception quando da erro? criar uma personalizada.

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,65 @@
+
+            HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(formData, headers);
+
+            ResponseEntity<String> response = restTemplate.postForEntity(
```
pq o retorno é void e voce ta definindo uma variável?
Faz um retorno boolean true ou false, valido ou nao.
se retornar o token, vc da um return true, se nao, return false, se for false, deve ter uma validação na controller que retorna um bad request

paulopraddo replied:
Não faz sentido a função retornar um boolean de true ou false pq o resttemplate lança uma exceção sempre que tem uma resposta de 400 pra cima, só defini para podermos utilizar a frente no projeto.

Bighetto commented on src/main/java/api/config/V8BankConfig.java:
```diff
@@ -0,0 +1,28 @@
+    private String clientId;
+
+    private String grantType = "password";
+    private String scope = "offline_access";
```
grantType e scope devem ser envs tambem
o valor sempre vai ser o mesmo, porem nao pode ficar fincado na aplicação isso, viola a segurança

Bighetto commented on src/main/resources/application.properties:
```diff
@@ -9,6 +9,9 @@ spring.jpa.properties.hibernate.dialect =org.hibernate.dialect.MySQL8Dialect
 api.security.token.secret=${API_SECRET_TOKEN}
+api.v8.bank.url=${V8_BANK_URL_GET_TOKEN}
+api.v8.bank.audience=${V8_BANK_AUDIENCE}
+api.v8.bank.clientid=${V8_BANK_CLIENT_ID}
```
adiciona as envs e valores no comentario do PR por gentileza, assim consigo atualizar no meu projeto tambem 
pode adicionar assim no comentario do PR
V8_BANK_URL_GET_TOKEN=hsaiduhsaidhsaiduhasi

paulopraddo committed:
Refac upload bank account

Envs:

V8_BANK_URL_GET_TOKEN=https://auth.v8sistema.com/oauth/token
V8_BANK_AUDIENCE=https://bff.v8sistema.com
V8_BANK_CLIENT_ID=DHWogdaYmEI8n5bwwxPDzulMlSK7dwIn
V8_BANK_GRANTTYPE=password
V8_BANK_SCOPE=offline_access

Constraint para que não seja possivel inserir bank users com informações de login repetidas:

ALTER TABLE bank_user
ADD CONSTRAINT unique_bank_login
UNIQUE KEY (bank_id, login);

Bighetto commented on src/main/java/api/bank/app/entrypoints/BankController.java:
```diff
@@ -24,6 +32,30 @@ public class BankController implements BankResource {
+            return ResponseEntity.ok().build();
+
+        } catch (UserBankV8ValidationException e) {
+            return ResponseEntity.badRequest().body(e.getMessage());
```
no body aqui voce deve colocar mensagens explicitas, no caso se o catch foi de UserBankV8ValidationException, retorna no body o seguinte :
Usuario de banco invalido, juntamente com o bad request.
Isso vai servir pra identificar melhor o erro no front.
o e.getMessage() basicamente é o stacktrace da parada, sacou?

Bighetto commented on src/main/java/api/bank/app/entrypoints/BankController.java:
```diff
@@ -24,6 +32,30 @@ public class BankController implements BankResource {
+        } catch (UserBankV8ValidationException e) {
+            return ResponseEntity.badRequest().body(e.getMessage());
+        } catch (BankUserAlreadyExistsException e) {
+            return ResponseEntity.badRequest().body(e.getMessage());
```
mesma coisa aqui, defina uma mensagem de retorno no endpoint, neste caso pode ser : Usuario de banco ja existente. e um bad request, pq eu vou validar no front pela mensagem, alem do bad request

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,80 @@
+            v8BankRequestDTO.setGrantType(this.v8BankDataProvider.getGrantType());
+            v8BankRequestDTO.setScope(this.v8BankDataProvider.getScope());
+
+            System.out.println("Username: " + v8BankRequestDTO.getUsername());
```
remover system out do codigo, deixar apenas lógica

Bighetto commented on src/main/java/api/bank/domain/service/ValidateUserBankV8Service.java:
```diff
@@ -0,0 +1,80 @@
+
+            HttpEntity<MultiValueMap<String, String>> request = new HttpEntity<>(formData, headers);
+
+            restTemplate.postForEntity(
```
como comentei, deve recuperar o valor dessa requisição, validar se no response tem o acess_token preenchido corretamente, se tiver, retorna um true, se nao tiver, pode retornar um false, e isso vc vai retornar usuario valido ou invalido, antes de salvar o mesmo.

Bighetto reviewed (changes requested):
realizar correções e solicitar review novamente

paulopraddo committed:
refac exception message
